
# Homework 4


Some initializations

In [1]:
import numpy as np
import pandas as pd
import re
import matplotlib.pyplot as plt
import seaborn as sns
import sqlite3 as sql
pd.set_option('display.max_columns', 62) # to see all columns 

pd.options.mode.copy_on_write = True # to avoid warnings when replacing values in a column and other commands
# see https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html#copy-on-write-chained-assignment


## Joining data



#### Loading data


In [2]:
# ![Upper left corner of the file opened in Jupyter](cleared_cell_phones_total-csv.jpg "'Delimiter' shows the sep parameter value to use")<br><br>

In [3]:
cleaned_cell_phones_total_df = pd.read_csv('cleaned_cell_phones_total.csv', index_col='iso-3')

In [4]:
# cleared_cell_phones_total_data_df = pd.read_csv('cleared_cell_phones_total.csv', index_col='iso-3')

In [5]:
# ![Upper left corner of the file opened in Jupyter](country_data-csv.jpg "'Delimiter' shows the sep parameter value to use")<br><br>

In [6]:
country_data_df = pd.read_csv("country_data.csv", index_col='alpha-3')

In [7]:
# country_data_df = pd.read_csv("https://github.com/su-mt4007/data/country_data.csv", index_col='alpha-3')

In [8]:
# ![Upper left corner of the file opened in Jupyter](pop_data-csv.jpg "'Delimiter' shows the sep parameter value to use")<br><br>

In [9]:
pop_data_df = pd.read_csv('pop_data.csv', index_col='iso-3')

In [10]:
# pop_data_df = pd.read_csv('https://github.com/su-mt4007/data/pop_data.csv', index_col='iso-3')


#### Structure of data


Cell phones

In [11]:
cleaned_cell_phones_total_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 214 entries, ABW to ZWE
Data columns (total 56 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   1960    214 non-null    float64
 1   1965    214 non-null    float64
 2   1966    214 non-null    float64
 3   1967    214 non-null    float64
 4   1968    214 non-null    float64
 5   1969    214 non-null    float64
 6   1970    214 non-null    float64
 7   1971    214 non-null    float64
 8   1972    214 non-null    float64
 9   1973    214 non-null    float64
 10  1974    214 non-null    float64
 11  1975    214 non-null    float64
 12  1976    214 non-null    float64
 13  1977    214 non-null    float64
 14  1978    214 non-null    float64
 15  1979    214 non-null    float64
 16  1980    214 non-null    float64
 17  1981    214 non-null    float64
 18  1982    214 non-null    float64
 19  1983    214 non-null    float64
 20  1984    214 non-null    float64
 21  1985    214 non-null    float64
 22  1986 

In [12]:
cleaned_cell_phones_total_df.index.name

'iso-3'

Countries

In [13]:
country_data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 249 entries, AFG to ZWE
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   name                      249 non-null    object 
 1   alpha-2                   248 non-null    object 
 2   country-code              249 non-null    int64  
 3   iso_3166-2                249 non-null    object 
 4   region                    248 non-null    object 
 5   sub-region                248 non-null    object 
 6   intermediate-region       107 non-null    object 
 7   region-code               248 non-null    float64
 8   sub-region-code           248 non-null    float64
 9   intermediate-region-code  107 non-null    float64
dtypes: float64(3), int64(1), object(6)
memory usage: 21.4+ KB


In [14]:
country_data_df.index.name

'alpha-3'

In [15]:
print(country_data_df.duplicated().sum()) # are there any duplicates

0


Some more data

In [16]:
country_data_df.head()

name alpha-2  country-code     iso_3166-2   region  \
alpha-3                                                                 
AFG         Afghanistan      AF             4  ISO 3166-2:AF     Asia   
ALA       Åland Islands      AX           248  ISO 3166-2:AX   Europe   
ALB             Albania      AL             8  ISO 3166-2:AL   Europe   
DZA             Algeria      DZ            12  ISO 3166-2:DZ   Africa   
ASM      American Samoa      AS            16  ISO 3166-2:AS  Oceania   

              sub-region intermediate-region  region-code  sub-region-code  \
alpha-3                                                                      
AFG        Southern Asia                 NaN        142.0             34.0   
ALA      Northern Europe                 NaN        150.0            154.0   
ALB      Southern Europe                 NaN        150.0             39.0   
DZA      Northern Africa                 NaN          2.0             15.0   
ASM            Polynesia                 NaN          9.0             61.0   

         intermediate-region-code  
alpha-3                            
AFG                           NaN  
ALA                           NaN  
ALB                           NaN  
DZA                           NaN  
ASM                           NaN

Populations

In [17]:
pop_data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 266 entries, ABW to ZWE
Data columns (total 62 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  266 non-null    int64  
 1   1960        264 non-null    float64
 2   1961        264 non-null    float64
 3   1962        264 non-null    float64
 4   1963        264 non-null    float64
 5   1964        264 non-null    float64
 6   1965        264 non-null    float64
 7   1966        264 non-null    float64
 8   1967        264 non-null    float64
 9   1968        264 non-null    float64
 10  1969        264 non-null    float64
 11  1970        264 non-null    float64
 12  1971        264 non-null    float64
 13  1972        264 non-null    float64
 14  1973        264 non-null    float64
 15  1974        264 non-null    float64
 16  1975        264 non-null    float64
 17  1976        264 non-null    float64
 18  1977        264 non-null    float64
 19  1978        264 non-null    floa

In [18]:
pop_data_df.index.name

'iso-3'

In [19]:
print(pop_data_df.duplicated().sum())

0


Some more data

In [20]:
pop_data_df.tail()

Unnamed: 0        1960        1961        1962        1963        1964  \
iso-3                                                                           
XKX           261    947000.0    966000.0    994000.0   1022000.0   1050000.0   
YEM           262   5542459.0   5646668.0   5753386.0   5860197.0   5973803.0   
ZAF           263  16520441.0  16989464.0  17503133.0  18042215.0  18603097.0   
ZMB           264   3119430.0   3219451.0   3323427.0   3431381.0   3542764.0   
ZWE           265   3806310.0   3925952.0   4049778.0   4177931.0   4310332.0   

             1965        1966        1967        1968        1969        1970  \
iso-3                                                                           
XKX     1078000.0   1106000.0   1135000.0   1163000.0   1191000.0   1219000.0   
YEM     6097298.0   6228430.0   6368014.0   6515904.0   6673981.0   6843607.0   
ZAF    19187194.0  19789771.0  20410677.0  21050540.0  21704214.0  22368306.0   
ZMB     3658024.0   3777680.0   3901288.0   4029173.0   4159007.0   4281671.0   
ZWE     4447149.0   4588529.0   4734694.0   4886347.0   5044163.0   5202918.0   

             1971        1972        1973        1974        1975        1976  \
iso-3                                                                           
XKX     1247000.0   1278000.0   1308000.0   1339000.0   1369000.0   1400000.0   
YEM     7024196.0   7215835.0   7417736.0   7630190.0   7855657.0   8094985.0   
ZAF    23031441.0  23698507.0  24382513.0  25077016.0  25777964.0  26480300.0   
ZMB     4399919.0   4523581.0   4653289.0   4789038.0   4931249.0   5079672.0   
ZWE     5363423.0   5532842.0   5712712.0   5903530.0   6097083.0   6288387.0   

             1977        1978        1979        1980        1981        1982  \
iso-3                                                                           
XKX     1430000.0   1460000.0   1491000.0   1521000.0   1552000.0   1582000.0   
YEM     8348182.0   8615301.0   8899922.0   9204938.0   9529105.0   9872292.0   
ZAF    27199838.0  27943445.0  28697014.0  29463549.0  30232561.0  31022417.0   
ZMB     5233292.0   5391355.0   5553462.0   5720438.0   5897481.0   6090818.0   
ZWE     6453044.0   6549349.0   6655833.0   7049926.0   7506526.0   7803855.0   

             1983        1984        1985        1986        1987        1988  \
iso-3                                                                           
XKX     1614000.0   1647000.0   1682000.0   1717000.0   1753000.0   1791000.0   
YEM    10237391.0  10625687.0  11036918.0  11465444.0  11915563.0  12387238.0   
ZAF    31865176.0  32768207.0  33752964.0  34877834.0  36119333.0  37393853.0   
ZMB     6291070.0   6488072.0   6686449.0   6890967.0   7095185.0   7294325.0   
ZWE     8106356.0   8398567.0   8690515.0   8983044.0   9277488.0   9568739.0   

             1989        1990        1991        1992        1993        1994  \
iso-3                                                                           
XKX     1827000.0   1862000.0   1898000.0   1932000.0   1965000.0   1997000.0   
YEM    12872362.0  13375121.0  13895851.0  14433771.0  14988047.0  15553171.0   
ZAF    38668684.0  39877570.0  40910959.0  41760755.0  42525440.0  43267982.0   
ZMB     7491275.0   7686401.0   7880466.0   8074337.0   8270917.0   8474216.0   
ZWE     9846346.0  10113893.0  10377815.0  10641501.0  10794918.0  10858594.0   

             1995        1996        1997        1998        1999        2000  \
iso-3                                                                           
XKX     2029000.0   2059000.0   2086000.0   1966000.0   1762000.0   1700000.0   
YEM    16103339.0  16614326.0  17108681.0  17608133.0  18114552.0  18628700.0   
ZAF    43986084.0  44661603.0  45285048.0  45852166.0  46364681.0  46813266.0   
ZMB     8684135.0   8902019.0   9133156.0   9372430.0   9621238.0   9891136.0   
ZWE    10994041.0  11178171.0  11362401.0  11548364.0  11716454.0  11834676.0   

             2001        2002  


#### Cleaning data


We know that the data in **cleaned_cell_phones_total_df** is cleaned already.

In **country_data_df** there are many null values. To save a lot of work and time, it would be good to only clean data that will be needed to solve the assignment.

It seems that only the index **alpha-3** and **name** column will be used. The column has no null values, but its Dtype is object. Converting this to Dtype string would be beneficial.

First, make a deep copy of country_data_df.

In [21]:
name_strings_country_data_df = country_data_df.copy()

Now, convert the values in the name column to strings.

In [22]:
name_strings_country_data_df['name'] = country_data_df['name'].astype("string")

In [23]:
name_strings_country_data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 249 entries, AFG to ZWE
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   name                      249 non-null    string 
 1   alpha-2                   248 non-null    object 
 2   country-code              249 non-null    int64  
 3   iso_3166-2                249 non-null    object 
 4   region                    248 non-null    object 
 5   sub-region                248 non-null    object 
 6   intermediate-region       107 non-null    object 
 7   region-code               248 non-null    float64
 8   sub-region-code           248 non-null    float64
 9   intermediate-region-code  107 non-null    float64
dtypes: float64(3), int64(1), object(5), string(1)
memory usage: 21.4+ KB


In **pop_data_df** there is a column we don´t need, so we drop it.

In [24]:
# pop_data_df.columns

In [25]:
dropped_pop_data_df = pop_data_df.drop(columns='Unnamed: 0')

In [26]:
# dropped_pop_data_df.info()


#### Merge dataframes


To create the first table in the assignment we can join, merge, cleaned_cell_phones_total_df and pop_data_df. Since there are fewer countries 
in cleaned_cell_phones_total_df, 214, than in pop_data_df, 266, we will make an inner join that only includes the countries where we have data 
both on the number of cellphones and the sizes of the populations. 

There are also fewer years in cleaned_cell_phones_total_df, 56, than in dropped_pop_data_df, 61, which also motivates an inner join.

For every year in pop_data_df there is one or two countries that have no value. It is first after the join that we can find out if some of the
included countries is affected by missing values.

To get the right format for the first table we have to make multi-indices, 'iso-3' and 'year', on both dataframes.

In [27]:
stacked_cleaned_cell_phones_total_df = cleaned_cell_phones_total_df.stack()

In [28]:
stacked_cleaned_cell_phones_total_df.info()

<class 'pandas.core.series.Series'>
MultiIndex: 11984 entries, ('ABW', '1960') to ('ZWE', '2019')
Series name: None
Non-Null Count  Dtype  
--------------  -----  
11984 non-null  float64
dtypes: float64(1)
memory usage: 141.1+ KB


In [29]:
# stacked_cleaned_cell_phones_total_df.index.names

In [30]:
stacked_cleaned_cell_phones_total_df.index.names = ['iso-3', 'year']

In [31]:
# stacked_cleaned_cell_phones_total_df.columns.name = ['n_cellphones']

In [32]:
stacked_cleaned_cell_phones_total_df.head()

iso-3  year
ABW    1960    0.0
       1965    0.0
       1966    0.0
       1967    0.0
       1968    0.0
dtype: float64

In [33]:
stacked_cleaned_cell_phones_total_df.tail()

iso-3  year
ZWE    2015    12800000.0
       2016    12900000.0
       2017    14100000.0
       2018    12900000.0
       2019    13200000.0
dtype: float64

In [34]:
stacked_dropped_pop_data_df = dropped_pop_data_df.stack()

In [35]:
stacked_dropped_pop_data_df.info()

<class 'pandas.core.series.Series'>
MultiIndex: 16135 entries, ('ABW', '1960') to ('ZWE', '2020')
Series name: None
Non-Null Count  Dtype  
--------------  -----  
16135 non-null  float64
dtypes: float64(1)
memory usage: 186.1+ KB


In [36]:
stacked_dropped_pop_data_df.index.names = ['iso-3', 'year']

In [37]:
stacked_dropped_pop_data_df.index.names

FrozenList(['iso-3', 'year'])

In [38]:
stacked_dropped_pop_data_df.head()

iso-3  year
ABW    1960    54608.0
       1961    55811.0
       1962    56682.0
       1963    57475.0
       1964    58178.0
dtype: float64

In [39]:
stacked_dropped_pop_data_df.tail()

iso-3  year
ZWE    2016    14452704.0
       2017    14751101.0
       2018    15052184.0
       2019    15354608.0
       2020    15669666.0
dtype: float64

Now we can merge the two series to a dataframe. They must be named to be able to merge.

In [40]:
stacked_cleaned_cell_phones_total_df.name = 'n_cellphones'

In [41]:
stacked_cleaned_cell_phones_total_df.info()

<class 'pandas.core.series.Series'>
MultiIndex: 11984 entries, ('ABW', '1960') to ('ZWE', '2019')
Series name: n_cellphones
Non-Null Count  Dtype  
--------------  -----  
11984 non-null  float64
dtypes: float64(1)
memory usage: 141.1+ KB


In [42]:
stacked_dropped_pop_data_df.name = 'population'

In [43]:
stacked_dropped_pop_data_df.info()

<class 'pandas.core.series.Series'>
MultiIndex: 16135 entries, ('ABW', '1960') to ('ZWE', '2020')
Series name: population
Non-Null Count  Dtype  
--------------  -----  
16135 non-null  float64
dtypes: float64(1)
memory usage: 186.1+ KB


In [44]:
# table_1 = stacked_cleaned_cell_phones_total_df.merge(stacked_dropped_pop_data_df, how='inner', on=['iso-3', 'year'], )

In [45]:
table_1 = pd.merge(left=stacked_cleaned_cell_phones_total_df, right=stacked_dropped_pop_data_df, how='inner', on=['iso-3', 'year'])

In [46]:
table_1.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 11958 entries, ('ABW', '1960') to ('ZWE', '2019')
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   n_cellphones  11958 non-null  float64
 1   population    11958 non-null  float64
dtypes: float64(2)
memory usage: 234.2+ KB


In [47]:
table_1.head()

n_cellphones  population
iso-3 year                          
ABW   1960           0.0     54608.0
      1965           0.0     58782.0
      1966           0.0     59291.0
      1967           0.0     59522.0
      1968           0.0     59471.0

In [48]:
table_1.tail()

n_cellphones  population
iso-3 year                          
ZWE   2015    12800000.0  14154937.0
      2016    12900000.0  14452704.0
      2017    14100000.0  14751101.0
      2018    12900000.0  15052184.0
      2019    13200000.0  15354608.0

In [49]:
resetted_table_1 = table_1.reset_index()

In [50]:
resetted_table_1.head()

iso-3  year  n_cellphones  population
0   ABW  1960           0.0     54608.0
1   ABW  1965           0.0     58782.0
2   ABW  1966           0.0     59291.0
3   ABW  1967           0.0     59522.0
4   ABW  1968           0.0     59471.0

In [51]:
sorted_resetted_table_1 = resetted_table_1.sort_values(by=['year', 'iso-3'])

In [52]:
sorted_resetted_table_1.head()

iso-3  year  n_cellphones  population
0     ABW  1960           0.0     54608.0
56    AFG  1960           0.0   8622466.0
112   AGO  1960           0.0   5357195.0
168   ALB  1960           0.0   1608800.0
224   AND  1960           0.0      9443.0

In [53]:
sorted_resetted_table_1.tail()

iso-3  year  n_cellphones  population
11733   XKX  2019      562000.0   1788878.0
11789   YEM  2019    15300000.0  31546691.0
11845   ZAF  2019    97000000.0  58087055.0
11901   ZMB  2019    17200000.0  18380477.0
11957   ZWE  2019    13200000.0  15354608.0

In [54]:
styled_sorted_resetted_table_3 = sorted_resetted_table_1.iloc[:5,:].style\
                        .format(formatter='{:.4g}'.format())\
                        .hide(axis='index')

IndexError: Replacement index 0 out of range for positional args tuple

In [ ]:
styled_sorted_resetted_table_3

In [ ]:
styled_sorted_resetted_table_1 = sorted_resetted_table_1.iloc[:10,:].style\
                        .hide(axis='index')

In [ ]:
styled_sorted_resetted_table_1

In [ ]:
styled_sorted_resetted_table_2 = sorted_resetted_table_1.iloc[-10:,:].style\
                        .hide(axis='index')

In [ ]:
styled_sorted_resetted_table_2


#### Old code, remove ??


For **pop_data_df** it seems that we only need data for the years 1960 and 2010-2019, for a few countries. These countries are ABW, AFG, AGO, ALB, AND, Macao, Hong Kong, United Arab Emirates,	Antigua and Barbuda, and Seychelles.

There are some null values in pop_data_df. We have to check if any of these concerns the years and countries we will treat.

First, we will get the alpha-3 codes for the names.

In [ ]:
alpha_3_codes = name_strings_country_data_df['name'].isin(['Macao', 'Hong Kong', 'United Arab Emirates','Antigua and Barbuda', 'Seychelles'])

In [ ]:
name_strings_country_data_df[alpha_3_codes]

Now we can select the interesting rows in pop_data_df.

In [ ]:
interesting_countries_bool = pop_data_df.index.isin(['ATG','HKG','MAC','SYC','ARE','ABW', 'AFG', 'AGO', 'ALB', 'AND'])

In [ ]:
interesting_years = ['1960', '2010','2011','2012','2013','2014','2015','2016','2017','2018','2019']

Make a deep copy of pop_data_df and then choose the interesting countries

In [ ]:
interesting_countries_df = pop_data_df[interesting_countries_bool].copy()

In [ ]:
interesting_countries_df.info()

Make a deep copy of interesting_countries_df

In [ ]:
interesting_countries_and_years_df = interesting_countries_df[interesting_years].copy()

In [ ]:
interesting_countries_and_years_df.info()

In [ ]:
interesting_countries_and_years_df

Set a name for all the columns

In [ ]:
interesting_countries_and_years_df.columns.name = 'year'

In [ ]:
interesting_countries_and_years_df

Tests

In [ ]:
test_stack_df = interesting_countries_and_years_df.stack()

In [ ]:
test_stack_df

In [ ]:
type(test_stack_df)

In [ ]:
test_stack_df.index


##### We see that there are no null values to handle for the countries and years we are interested in.



##### Handle cleaned_cell_phones_total_data_df the same way


We will choose the same countries and years as above, for the data on cell phones.

In [ ]:
interesting_countries_cell_phones_bool = cleaned_cell_phones_total_data_df.index.isin(['ATG','HKG','MAC','SYC','ARE','ABW', 'AFG', 'AGO', 'ALB', 'AND'])

In [ ]:
interesting_countries_cell_phones_df = cleaned_cell_phones_total_data_df[interesting_countries_cell_phones_bool].copy()

In [ ]:
interesting_countries_cell_phones_df.info()

In [ ]:
interesting_countries_and_years_cell_phones_df = interesting_countries_cell_phones_df[interesting_years].copy()

In [ ]:
interesting_countries_and_years_cell_phones_df.columns

In [ ]:
interesting_countries_and_years_cell_phones_df.columns.name = 'year'

In [ ]:
interesting_countries_and_years_cell_phones_df.columns

In [ ]:
interesting_countries_and_years_cell_phones_df.info()

In [ ]:
interesting_countries_and_years_cell_phones_df


#### Stacking and joining data frames


The data frames interesting_countries_and_years_df and interesting_countries_and_years_cell_phones_df are in wide format. To make it easier to join them on index and year, it would be good to stack them, to long format.


##### Stack interesting_countries_and_years_df


In [ ]:
stacked_interesting_countries_and_years_df = interesting_countries_and_years_df.stack()

In [ ]:
stacked_interesting_countries_and_years_df.info()

Give a name to the Series

In [ ]:
stacked_interesting_countries_and_years_df.name = 'population'

In [ ]:
stacked_interesting_countries_and_years_df.info()

In [ ]:
stacked_interesting_countries_and_years_df

In [ ]:
stacked_interesting_countries_and_years_df['ABW']


##### Stack interesting_countries_and_years_cell_phones_df


In [ ]:
stacked_interesting_countries_and_years_cell_phones_df = interesting_countries_and_years_cell_phones_df.stack()

In [ ]:
stacked_interesting_countries_and_years_cell_phones_df.info()

Give a name to the Series

In [ ]:
stacked_interesting_countries_and_years_cell_phones_df.name = 'n_cellphones'

In [ ]:
stacked_interesting_countries_and_years_cell_phones_df.info()

In [ ]:
stacked_interesting_countries_and_years_cell_phones_df

In [ ]:
stacked_interesting_countries_and_years_cell_phones_df['ABW']


##### Join stacked_interesting_countries_and_years_cell_phones_df and stacked_interesting_countries_and_years_df


Now we can join the two stacked data frames on the multiindex [iso-3, year]

It is easier to join data frames than Series, so we create Data Frames from the Series stacked_interesting_countries_and_years_cell_phones_df and stacked_interesting_countries_and_years_df (which have gotten misleading names).

In [ ]:
n_cellphones_df = pd.DataFrame(data=stacked_interesting_countries_and_years_cell_phones_df)

In [ ]:
n_cellphones_df

In [ ]:
population_df = pd.DataFrame(data=stacked_interesting_countries_and_years_df)

In [ ]:
population_df

In [ ]:
n_cellphones_and_population_df = n_cellphones_df.merge(population_df, on=['iso-3','year'] , copy=True)

In [ ]:
n_cellphones_and_population_df

In [ ]:
type(n_cellphones_and_population_df)

##### Now we can show the first table required in the assignment.

First we unstack the data frame two levels.

In [ ]:
unstacked_n_cellphones_and_population_df = n_cellphones_and_population_df.unstack(level=-1)

In [ ]:
unstacked_n_cellphones_and_population_df

In [ ]:
n_cellphones_and_population_df.index

In [ ]:
first_table_1960_df = n_cellphones_and_population_df.xs('1960', level = 'year')

In [ ]:
first_table_1960_df 

In [ ]:
first_table_1960_df.info()

In [ ]:
# df.loc[(slice(None), "one"), :]

Another subset

In [ ]:
first_table_1960_df_II = n_cellphones_and_population_df.loc[(slice(None), "1960"), :]


#### The first table


In [ ]:
first_table_1960_df_II.head()


##### The second table


We have to compute the number of cell phones per capita for the different countries. First a deep copy.

In [ ]:
n_cellphones_and_population_and_per_capita_df = n_cellphones_and_population_df.copy()

In [ ]:
n_cellphones_and_population_and_per_capita_df['phones_per_capita'] = n_cellphones_and_population_and_per_capita_df['n_cellphones'] / n_cellphones_and_population_and_per_capita_df['population'] 

In [ ]:
n_cellphones_and_population_and_per_capita_df.info()

In [ ]:
n_cellphones_and_population_and_per_capita_df

In the second table we are only interested in the column phones_per_capita, so we choose a subset.

In [ ]:
phones__per_capita_Series = n_cellphones_and_population_and_per_capita_df['phones_per_capita'].copy()

In [ ]:
phones__per_capita_Series.info()

This is a Series, so we convert it to a DataFrame.

In [ ]:
phones__per_capita_df = pd.DataFrame(data=phones__per_capita_Series)

In [ ]:
phones__per_capita_df.info()

In [ ]:
phones__per_capita_df

We want the country names and not the iso-3 codes, so we make a left join. First though, we make an unstack.

In [ ]:
unstacked_phones_per_capita_df = phones__per_capita_df.unstack(level=-1)

In [ ]:
unstacked_phones_per_capita_df.info()

In [ ]:
unstacked_phones_per_capita_df

In [ ]:
unstacked_phones_per_capita_df.columns

We will not need the column for 1960, so we delete it.

In [ ]:
years_unstacked_phones_per_capita_df = unstacked_phones_per_capita_df.drop(axis=1, columns=('phones_per_capita','1960'))

In [ ]:
years_unstacked_phones_per_capita_df

We are only interested in the last five countries.

In [ ]:
countries_and_years_unstacked_phones_per_capita_df = years_unstacked_phones_per_capita_df.tail().copy()

In [ ]:
countries_and_years_unstacked_phones_per_capita_df.info()


#### The second table


Time ran out to do more than this, for the first part of the assignment.

In [ ]:
countries_and_years_unstacked_phones_per_capita_df

Now we can choose the relevant column from name_strings_country_data_df

In [ ]:
relevant_name_strings_country_data_Series = name_strings_country_data_df['name'].copy()

In [ ]:
relevant_name_strings_country_data_Series.info()

Convert to data frame

In [ ]:
relevant_name_strings_country_data_df = pd.DataFrame(data=relevant_name_strings_country_data_Series)

In [ ]:
relevant_name_strings_country_data_df.info()

Time to join 


## SQL


Connect to the database and look at the data

In [ ]:
conn = sql.connect("./user_actions.db")
cursor = conn.cursor()

In [ ]:
tables = list(cursor.execute("SELECT name FROM sqlite_master WHERE type='table';"))
tables

In [ ]:
views = list(cursor.execute("SELECT name FROM sqlite_master WHERE type='view';"))
views

In [ ]:
# SELECT

query = """
SELECT * 
FROM user_actions
"""

cursor.execute(query)

for data in cursor.fetchone():    
    print(data)
user_action_pd = pd.read_sql_query(query, conn)

In [ ]:
user_action_pd.info()


#### Users who have "signedup"


In [ ]:
query = """
SELECT username 
FROM user_actions
WHERE action='signup'
"""

cursor.execute(query)

# for data in cursor.fetchall():    
#    print(data)

pd.read_sql_query(query, conn).sort_values(by='username')


#### Total number of log entries for each user


In [ ]:
query = """
SELECT user_id, username, count(*) 
FROM user_actions
WHERE action='login'
GROUP BY username
"""

cursor.execute(query)

# for data in cursor.fetchall():  
#     print(data)

pd.read_sql_query(query, conn).sort_values(by='username')


#### Users who have logged in and signed up on the same day


Create two views, for signup and login, with new column, 'day', from 'timestamp'.

In [ ]:
query = """
SELECT *
FROM user_actions
"""

#cursor.execute(query)

# for data in cursor.fetchall():
# for data in cursor.fetchone():    
#    print(data)

In [ ]:
query = """
CREATE VIEW login_view 
 (user_id, login_username, login_day)  
AS SELECT user_id, username, date(timestamp)
FROM user_actions
WHERE action='login' 
GROUP BY username, date(timestamp);
"""
cursor.execute(query);

#for data in cursor.fetchone():
# for data in cursor.fetchone():    
#    print(data)
# pd.read_sql_query(query, conn)
#pd.read_sql_query(query, conn).sort_values(by='username')

In [ ]:
query = """
CREATE VIEW signup_view 
  (user_id, signup_username, signup_day) 
AS SELECT user_id, username, date(timestamp)
FROM user_actions
WHERE action='signup' 
GROUP BY username, date(timestamp);
"""
cursor.execute(query);

# for data in cursor.fetchone():
# for data in cursor.fetchone():    
#    print(data)
# pd.read_sql_query(query, conn)
# pd.read_sql_query(query, conn).sort_values(by='username')


##### Users who have signed up and logged in the same day.


In [ ]:
query = """
SELECT signup_view.user_id, signup_username, signup_day
FROM signup_view, login_view
WHERE signup_username = login_username
      AND signup_day = login_day
GROUP BY login_username
"""
cursor.execute(query)
# for data in cursor.fetchone():
# for data in cursor.fetchone():    
#    print(data)
# pd.read_sql_query(query, conn)
pd.read_sql_query(query, conn).sort_values(by='signup_day')

Cleaning up by dropping the views.

In [ ]:
query = """
DROP VIEW login_view
"""
cursor.execute(query);
#AND login_view.date(timestamp) = signup_view.date(timestamp) 
# WHERE login_view.username = signup_view.username 
# GROUP BY username
#for data in cursor.fetchone():
# for data in cursor.fetchone():    
#   print(data)
# pd.read_sql_query(query, conn)
# pd.read_sql_query(query, conn).sort_values(by='username')

In [ ]:
query = """
DROP VIEW signup_view
"""
cursor.execute(query);

In [ ]:
query = """
SELECT user_id, username, action, date(timestamp)
FROM user_actions
WHERE username='user1' AND (action='login' OR action='signup')
GROUP BY date(timestamp)
"""
# cursor.execute(query)

# for data in cursor.fetchall():
# for data in cursor.fetchone():    
#    print(data)
# pd.read_sql_query(query, conn)
# pd.read_sql_query(query, conn).sort_values(by='username')

In [ ]:
query = """
SELECT user_id, username, day
FROM user_actions
WHERE action='login' OR action='signup'
GROUP BY day
"""

# cursor.execute(query)

# for data in cursor.fetchone():
# for data in cursor.fetchone():    
#     print(data)
# pd.read_sql_query(query, conn)
# pd.read_sql_query(query, conn).sort_values(by='username')


## Regex



#### Read data from file


In [ ]:
first_part = r"^\d{1,3}\." # split each line at the beginning number, which will not be saved

In [ ]:
comments_pd = pd.read_table('comments.txt', sep=first_part, names=['id', 'comment'], engine='python')

In [ ]:
comments_pd['comment'].head() # show the first five comments

In [ ]:
# comments_pd.info()

In [ ]:
# comments_pd.head()

In [ ]:
comments_pd.dtypes


##### Drop the 'id' column and convert the comments to strings


This will create a Series of strings.

In [ ]:
string_comments = comments_pd['comment'].astype({'comment' : 'string'}, copy=True)

In [ ]:
# string_comments.info()

In [ ]:
string_comments.head(6) # show the first six comments

In [ ]:
# string_comments.tail()

In [ ]:
# string_comments[99]

In [ ]:
# string_comments[97]

In [ ]:
# string_comments[1]


#### Solve tasks



##### All hashtags


In [ ]:
pattern_all_hashtags = r"[#]{1}[\w]+"
pattern_all_hashtags

In [ ]:
# tag_comment_1 = re.findall(pattern_all_hashtags, string_comments[1]) # the second comment
# print(tag_comment_1)


###### All hashtags are found in the first comment


In [ ]:
tag_comment_0 = re.findall(pattern_all_hashtags, string_comments[0])
print(tag_comment_0)


##### Both #programming and #python


In [ ]:
# pattern_both_test = r"[#programming]+ och [#python]+"
# pattern_both_test

In [ ]:
pattern_both = r"[#p]{1}[\w]+[o]{1}[\w]*[n]{1}[\w]*"
pattern_both

In [ ]:
# both_tag_comment_0 = re.findall(pattern_both, string_comments[0])
# print(both_tag_comment_0)


###### In the second comment, #python is found, but not #programming


In [ ]:
both_tag_comment_1 = re.findall(pattern_both, string_comments[1])
print(both_tag_comment_1)


###### To find both #python and #programming, we have to search in the sixth comment


In [ ]:
both_tag_comment_5 = re.findall(pattern_both, string_comments[5])
print(both_tag_comment_5)

We see that this patterna finds both #programming och #python, if they exist in a comment. It also finds some other tags, e.g. #coding, but it was not required in the assignment that no other tags should be found.


##### All unique hash tags


In [ ]:
all_tags = []
for i in range(99):
    all_tags.append(re.findall(pattern_all_hashtags, string_comments[i]))

In [ ]:
all_tags[:5]

In [ ]:
type(all_tags)

In [ ]:
#len(all_tags)

This is a list of lists. In python it is possible to transform it to a set, where all duplicates will disappear. As it is required to do everything with regex, I will not continue.


##### Från kurs_demo


In [ ]:
string = "Please contact support@example.com for assistance. For general inquiries, you can email info@company.com and @gmail.com"
print(string)

In [ ]:
pattern = r"\b[\w._%+-]+@[\w.-]+\.[A-Za-z]{2,4}\b"
r"\b[\w._%+-]+@[\w.-]+\.[A-Za-z]{2,4}\b"

In [ ]:
emails = re.findall(pattern, string)
print(emails)